### PR TITLE
REGR: DataFrame.agg with axis=1, EA dtype, and duplicate index

### LINKED ISSUE #42380: BUG: 1.3.0 DataFrame.agg over categorical columns with non-unique index returns wrong size result
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd

df = pd.DataFrame({"a": list("abcde"), "b": list("abcde")}, index=list("aabbc"), dtype="category")
df.agg("-".join, axis=1)
```

Using pandas 1.2.5:

```
a    a-a
a    b-b
b    c-c
b    d-d
c    e-e
dtype: object
```

Using pandas 1.3.0:

```
a    b-b
b    d-d
c    e-e
dtype: object
```

It does not look like this is an issue if I use `df.apply` instead of `df.agg`.

#### Problem description

When a aggregation of the rows is run on a dataframe with categorical columns and non-unique indices, the result is the wrong length.

It's weird that the output isn't the right length. Since I'm computing a value per row, I expect the same number of rows in the output as in the input. It's especially weird that this only happens if the columns are categorical.

That is:

```python
df = pd.DataFrame({"a": list("abcde"), "b": list("abcde")}, index=list("aabbc"))
df.agg("-".join, axis=1)
```

```
a    a-a
a    b-b
b    c-c
b    d-d
c    e-e
dtype: object
```

in both versions.

#### Expected Output

I would expect the same output between versions. The result given by 1.2.5 seems more correct to me at the moment.

#### Output of ``pd.show_versions()``

<details>

```
INSTALLED VERSIONS
------------------
commit           : f00ed8f47020034e752baf0250483053340971b0
python           : 3.8.10.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 20.5.0
Version          : Darwin Kernel Version 20.5.0: Sat May  8 05:10:33 PDT 2021; root:xnu-7195.121.3~9/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.0
numpy            : 1.21.0
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 21.1.3
setuptools       : 56.0.0
Cython           : 0.29.23
pytest           : 6.2.4
hypothesis       : None
sphinx           : 4.0.2
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.6.3
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.23.1
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : None
fsspec           : 2021.06.0
fastparquet      : 0.4.1
gcsfs            : None
matplotlib       : 3.4.2
numexpr          : 2.7.2
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 4.0.1
pyxlsb           : None
s3fs             : 0.4.2
scipy            : 1.7.0
sqlalchemy       : 1.3.18
tables           : 3.6.1
tabulate         : 0.8.7
xarray           : 0.18.2
xlrd             : 1.2.0
xlwt             : None
numba            : 0.53.1
```

</details>

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #42380
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

The issue in transpose was a bug that was introduced in 1.0.0, but induced a regression in 1.3.0 due to its use in `DataFrame.agg`.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -3338,12 +3338,10 @@ def transpose(self, *args, copy: bool = False) -> DataFrame:
         ):
             # We have EAs with the same dtype. We can preserve that dtype in transpose.
             dtype = dtypes[0]
-            arr_type = dtype.construct_array_type()
             values = self.values
 
-            new_values = [arr_type._from_sequence(row, dtype=dtype) for row in values]
             result = self._constructor(
-                dict(zip(self.index, new_values)), index=self.columns
+                values.T, index=self.columns, columns=self.index, dtype=dtype
```
Comment (jbrockmendel) [inline pandas/core/frame.py:3344]: values.T is going to cast to ndarray, which often means object

i think could use DataFrame.from_arrays?

Comment (rhshadrach) [reply pandas/core/frame.py:3344]: I see a MultiIndex.from_arrays, but not DataFrame. Here are some timings:

```
size = 1000
df = pd.DataFrame({'a': size * [1, 2, 3]}, dtype='category')
df_T = df.T

%timeit df.T
%timeit df_T.T
```

PR:

```
249 ms ± 6.39 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
92.3 ms ± 6.51 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Master:

```
237 ms ± 4.78 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
89.7 ms ± 6.87 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

so certainly a slight regression. Will continue looking to see if avoiding object is possible.

Comment (rhshadrach) [reply pandas/core/frame.py:3344]: Ah, found it - `_from_arrays`.

Comment (rhshadrach) [reply pandas/core/frame.py]: New timing with the update

```
234 ms ± 4.46 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
85 ms ± 4.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by jreback** [comment]
thanks @rhshadrach

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Comment by simonjayhawkins** [comment]
> The issue in transpose was a bug that was introduced in 1.0.0

in pandas 0.25.3 the returned DataFrame had object dtype.

```
>>> pd.__version__
'0.25.3'
>>> 
>>> df = pd.DataFrame(
...     {"a": list("abcde"), "b": list("abcde")}, index=list("aabbc"), dtype="category"
... )
>>> 
>>> df.T.dtypes
a    object
a    object
b    object
b    object
c    object
dtype: object
>>> 
```
on master, we now have value dependent behaviour

```
>>> pd.__version__
'1.4.0.dev0+193.g87d785599e'
>>> 
>>> df = pd.DataFrame(
...     {"a": list("abcde"), "b": list("abcde")}, index=list("aabbc"), dtype="category"
... )
>>> 
>>> df.T.dtypes
a    category
a    category
b    category
b    category
c    category
dtype: object
>>> 
>>> df = pd.DataFrame(
...     {"a": list("abcde"), "b": list("fghij")}, index=list("aabbc"), dtype="category"
... )
>>> 
>>> df.T.dtypes
a    object
a    object
b    object
b    object
c    object
dtype: object
>>> 
```

**Comment by rhshadrach** [comment]
@simonjayhawkins I think the column dtypes, not values, are what determines the resulting dtype. As such, I would not describe this as value-dependent behavior. The result looks like the correct one to me, but can open an issue if you would like this discussed further.

**Comment by simonjayhawkins** [comment]
> The result looks like the correct one to me

The result is now certainly consistent with the result for a unique index. (which also changed behavior between 0.25.3 and 1.0.0)

> but can open an issue if you would like this discussed further.

this has been backported for a patch release. and we could release anytime dependent on number and severity of reported regression.

My preference is to not having an open issue, but to keep the 1.3.x ready for release.

Any reason not just to revert #40428 instead?

**Comment by rhshadrach** [comment]
> this has been backported for a patch release. and we could release anytime dependent on number and severity of reported regression.
> 
> My preference is to not having an open issue, but to keep the 1.3.x ready for release.

Are you saying we should not fix the transpose regression from 0.25.3 -> 1.0.0 in 1.3.1? If that's the case, then I would support reverting #40428, assuming it does fix the issue in `agg`. However, the code prior to #40428 used transpose as well, so I don't understand how it could fix it (but might be missing something).

**Comment by simonjayhawkins** [comment]
> Are you saying we should not fix the transpose regression from 0.25.3 -> 1.0.0 in 1.3.1?

I am not saying that. I basically just wanted clarification of why #40428 was not reverted. I assume that because the root cause was determined as the change in #30091, that was fixed instead?

However, that change does not revert to the behavior in 0.25.3 because of the dtype change. (but that is also reasonable as the new behavior matches that of a unique index)

> If that's the case, then I would support reverting #40428, assuming it does fix the issue in `agg`. 

it is ok to fix prior regressions or bug fixes in patch releases, so that alone is not a reason to prefer the revert option.

> However, the code prior to #40428 used transpose as well, so I don't understand how it could fix it (but might be missing something).

that is odd.

**Comment by rhshadrach** [comment]
Thanks @simonjayhawkins, your assessment is correct - the root cause of the agg regression was the transpose regression. Both are tested in this PR.

> However, that change does not revert to the behavior in 0.25.3 because of the dtype change. (but that is also reasonable as the new behavior matches that of a unique index)

The behavior in 0.25.3 is undesirable. When a DataFrame has a single EA dtype, 0.25.3's transpose would convert it to object whereas it should remain the EA dtype. #30091 fixed this, but at the expense of silently dropping data when the index has duplicates. This PR maintains the desired behavior of #30091, but in a way that does not drop data when the index has duplicates.

**Pull request merged**